src/snarky/parties-leaves.ts

```diff
@@ -156,23 +161,207 @@ function toFields<K extends keyof TypeMap>(typeName: K, value: TypeMap[K]) {
   return ToFields[typeName](value);
 }
 
-let toJsonLeafTypes = new Set(Object.keys(ToJson));
-let toFieldsLeafTypes = new Set(Object.keys(ToFields));
+// to auxiliary
 
-// converters for types which got an annotation about its circuit type in Ocaml
+type ToAuxiliary = {
+  [K in keyof TypeMap]:
+    | ((x: TypeMap[K] | undefined) => [])
+    | ((x: TypeMap[K] | undefined) => [TypeMap[K]]);
+};
 
-type DataAsHash<T> = { data: T; hash: Field };
+let ToAuxiliary: ToAuxiliary = {
+  PublicKey: empty,
+  Field: empty,
+  Bool: empty,
+  AuthRequired: empty,
+  UInt32: empty,
+  UInt64: empty,
+  TokenId: empty,
+  Sign: empty,
+  // builtin
+  number: (x = 0) => [x],
+  null: empty,
+  undefined: empty,
+  string: (x = '') => [x],
+};
+
+function toAuxiliary<K extends keyof TypeMap>(
+  typeName: K,
+  value: TypeMap[K] | undefined
+) {
+  if (!(typeName in ToFields))
+    throw Error(`toAuxiliary: unsupported type "${typeName}"`);
+  return ToAuxiliary[typeName](value);
+}
+
+// size in fields
+
+type SizeInFields = { [K in keyof TypeMap]: number };
+let SizeInFields: SizeInFields = {
+  PublicKey: 2,
+  Field: 1,
+  Bool: 1,
+  AuthRequired: 3,
+  UInt32: 1,
+  UInt64: 1,
+  TokenId: 1,
+  Sign: 1,
+  // builtin
+  number: 0,
+  null: 0,
+  undefined: 0,
+  string: 0,
+};
 
-function convertEventsToJson({ data }: DataAsHash<Field[][]>) {
-  return data.map((row) => row.map((e) => toJson('Field', e)));
+function sizeInFields<K extends keyof TypeMap>(typeName: K) {
+  if (!(typeName in ToFields))
+    throw Error(`sizeInFields: unsupported type "${typeName}"`);
+  return SizeInFields[typeName];
 }
-function convertEventsToFields({ hash }: DataAsHash<Field[][]>) {
-  return [hash];
+
+// from fields & aux
+// these functions get the reversed output of `toFields` and `toAuxiliary` and pop the values they need from those arrays
+
+type FromFields = {
+  [K in keyof TypeMap]: (fields: Field[], aux: any[]) => TypeMap[K];
+};
+
+function takeOneAux(_: Field[], aux: any[]) {
+  return aux.pop()!;
 }
 
-function convertStringWithHashToJson({ data }: DataAsHash<string>) {
-  return data;
+let FromFields: FromFields = {
+  PublicKey(fields: Field[]) {
+    let x = fields.pop()!;
+    let isOdd = fields.pop()!;
+    // compute y from elliptic curve equation y^2 = x^3 + 5
+    // TODO: this is used in-snark, so we should improve constraint efficiency
+    let someY = x.mul(x).mul(x).add(5).sqrt();
+    let isTheRightY = isOdd.equals(someY.toBits()[0].toField());
+    let y = isTheRightY
+      .toField()
+      .mul(someY)
+      .add(isTheRightY.not().toField().mul(someY.neg()));
+    return new PublicKey(new Group(x, y));
+  },
+  Field(fields: Field[]) {
+    return fields.pop()!;
+  },
+  Bool(fields: Field[]) {
+    return Bool.Unsafe.ofField(fields.pop()!);
```

Comment:
this is only invoked on generated code we control from the json layout right?

Reply:
Yes! So it's safe because we know the field we're reading out was a Bool originally